# bed2_summary
get smallRNAseq summary in piclusters, transposons and genes from bed2 files generated by piPipes or other softwares
***
## installation
For easy install, run install.sh in *bed2_summary* folder after download and unzip the source code. It will add all the scripts needed into your PATH and PYTHONPATH. After installation, please use `source ~/.bashrc` or re-load the server.
After this, simple use `run_bed2_summary` to generate the summary of piPipes or mapping result.
***
## usage
*bed2_summary* need at least 4 input (first 4 parameters is required, others is optional):
1. -c control sample name with directory. Use without -t will only give out plots for control sample. eg: path_to_piPipes_result/sample_name_control
2. -o output directory. All the output files will be in this folder including bucket plots and summary. eg: results/piPipes/bed2_summary/
3. -g genome used. default: dm3
4. -n normalization method. default: miRNA 
    miRNA: normalized to reads per million mapped miRNA reads  
    uniq: normalized to reads per million mapped reads exclude miRNA and rRNA reads
5. -t treatment sample name with directory. If set, *bed2_summary* can make comparison between control and treatment
6. -G how deep you want to analysis genes. default: 1  
    0.) not analysis  
    1.) get normalized srna reads number and species for each gene  
    2.) also get buckets for each genes. It may takes more than 2 hour and the buckets pdf size may be more than 200M  
7. -p CPU numbers used in *bed2_summary*

tips:  
In *bed2_summary*, the input need to be **piPipes_output_folder/sample_name**.  
For example: To say if you used `piPipes small -i oreR_unox.cutadapt.fq.gz -o /project/common/piPipe.result/`, then the follow command is needed for bed2_summary:  
`run_bed2_summary -c  /project/common/piPipe.result/oreR_unox.cutadapt.fq.gz -o [which output folder you want to put the figures and summaries] -g dm3 -n [miRNA or uniq] [-G if you want to include gene analysis]`
Also, if you want to compare two conditions, like if you have ran piPipes for two conditions:  
`piPipes small -i oreR_unox.cutadapt.fq.gz -o /project/common/piPipe.result/`  
`piPipes small -i rhino_KO_unox.cutadapt.fq.gz -o /project/common/piPipe.result/`  
Then you can run:  
`run_bed2_summary -c  /project/common/piPipe.result/oreR_unox.cutadapt.fq.gz -t  /project/common/piPipe.result/rhino_KO_unox.cutadapt.fq.gz -o [which output folder you want to put the figures and summaries] -g dm3 -n [miRNA or uniq] [-G if you want to include gene analysis]`  
***
## output
*bed2_summary* can give three summary files: `prefix.picluster.summary`, `prefix.transposon.summary` and `prefix.gene.summary` which summarize informations for each picluster, transposon or gene in each row. And there are 10 columns in xxx.summary:
1. normalized sense+unique mapped reads.
2. normalized antisense+unique mapped reads.
3. normalized sense+all mapped reads.
4. normalized antisense+all mapped reads.
5. normalized sense+unique mapped small RNA species.
6. normalized antisense+unique mapped small RNA species.
7. normalized sense+all mapped small RNA species.
8. normalized antisense+all mapped small RNA species.
9. ping-pong zscore
10. normalized 10nt overlapped read pairs

*bed2_summary* can also output bucketplot, scatterplot and boxplot for piclusters, transposons and genes respectively. In the plot files, ping-pong score, length distribution and signal profile for each element is included.
![42AB](img/42AB.jpg "42AB")

![picluster](img/picluster.scatter.jpg "picluster")



## contact
please send questions or bugs to yutianxiong@gmail.com

## Prepare references for other organisms
```
assembly="mm10"
mkdir ~/tools/bed2_summary-6c01dc3/annotation/$assembly
samtools faidx ~/tools/piPipes-c93bde3/common/$assembly/$assembly.piRNAcluster.fa
cat ~/tools/piPipes-c93bde3/common/$assembly/$assembly.piRNAcluster.fa.fai | cut -f1,2 > ~/tools/bed2_summary-6c01dc3/annotation/$assembly/piRNAcluster.sizes
cat ~/tools/piPipes-c93bde3/common/$assembly/$assembly.genes.bed12 > ~/tools/bed2_summary-6c01dc3/annotation/$assembly/gene.bed12
samtools faidx ~/tools/piPipes-c93bde3/common/$assembly/$assembly.genes.fa
cat ~/tools/piPipes-c93bde3/common/$assembly/$assembly.genes.fa.fai | cut -f1,2 > ~/tools/bed2_summary-6c01dc3/annotation/$assembly/gene.sizes
samtools faidx ~/tools/piPipes-c93bde3/common/$assembly/$assembly.repBase.fa
cat ~/tools/piPipes-c93bde3/common/$assembly/$assembly.repBase.fa.fai | cut -f1,2 > ~/tools/bed2_summary-6c01dc3/annotation/$assembly/repBase.sizes
```

## Run example
After running piPipes, you can run:
### Single-sample mode

```
assembly="mm10"

name="sample1"
in_fastq="reads.1.adtrim.fastq.gz"
sdir=results/$name

mkdir -p $sdir/bed2_summary

run_bed2_summary \
    -c $sdir/${in_fastq%.fastq.gz} \
    -o $sdir/bed2_summary -g $assembly -n uniq -G 1 -p 1
```
### Dual-sample mode
```
assembly="mm10"

name1="sample1"
name2="sample2"

in_fastq="reads.1.adtrim.fastq.gz"

sdir1=$RES_DIR/$name1/piPipes
sdir2=$RES_DIR/$name2/piPipes
sdir=$RES_DIR/${name1}vs${name2}/piPipes

# Make shorter, nicer names
name1=`echo $name1 | sed 's/mmu.sRNASeq.//' | sed 's/\./-/g'`
name2=`echo $name2 | sed 's/mmu.sRNASeq.//' | sed 's/\./-/g'`
    
mkdir -p $sdir/bed2_summary
    
# Link original results so we have different input names
for a in transposon_piRNAcluster_mapping_normalized_by_allxmirna hairpins_mapping genome_mapping; do 
    echo $a
    mkdir -p $sdir/bed2_summary/$name1/$a
    mkdir -p $sdir/bed2_summary/$name2/$a
    
    for j in $sdir1/$a/*; do
        link=`echo $j | sed "s/${in_fastq%.fastq.gz}/$name1/"`
        ln -sf ../../../../../../$j $sdir/bed2_summary/$name1/$a/$(basename $link)
    done
    for j in $sdir2/$a/*; do
        link=`echo $j | sed "s/${in_fastq%.fastq.gz}/$name2/"`    
        ln -sf ../../../../../../$j $sdir/bed2_summary/$name2/$a/$(basename $link)
    done
done
    
run_bed2_summary \
    -c $sdir/bed2_summary/$name1/$name1 -t $sdir/bed2_summary/$name2/$name2 \
    -o $sdir/bed2_summary -g $assembly -n uniq -G 1 -p 1
```